components/datetime/src/format.rs

```diff
@@ -0,0 +1,55 @@
+use super::date::DateTime;
+use super::pattern::{parse_pattern, Field, FieldType};
+use std::fmt;
+
+pub struct FormattedDateTime<'s> {
+    pub(crate) pattern: &'s str,
```

Comment:
What is the benefit of returning the pattern as part of the `FormattedDateTime`?

Reply:
If the pattern is to be used to figure out what part of the string matches what field, this is probably not enough info.
For instance a pattern of "Md" can result in "123" and there is no way to tell if it is 1/23 or 12/3

There are also fields that don't show in the original pattern.
For example format date & time:
"Mar 12, 2020, at 9:32 am"
Fields:
date: "Mar 12, 2020
separator: ", at"
time: "9:32 am"
month: "Mar"
...
So there are overlapping fields (date = month + day + year, etc.)

Reply:
So the easiest thing is to create that field info as you format, and probably impossible to figure it out after the fact.

Reply:
> What is the benefit of returning the pattern as part of the FormattedDateTime?

I was looking for the performance/memory balance. I *think* we want to actually parse the pattern into fields in `format` and have `FormattedDateTime` already have pre-formatted fields, but that means I need to allocate a Vector for the fields, and performance/memory wise that is slower.

I still think it's the right thing to do, and expected that actual field computation will become heavier as we mature this, so eventually the tradeoff will pay off :)

Reply:
> There are also fields that don't show in the original pattern.

That's interesting. I don't think that's how `formatToParts` work in ECMA402. The way it works there is meant to allow you to do:

```js
let dateFormat = new Intl.DateTimeFormat('en');
let date = Date.now();

let formattedDate = dateFormat.formatToParts(date)
    .map(({value}) => value)
    .reduce((string, part) => string + part);

let fancyFormattedDate = dateFormat.formatToParts(date)
    .map(({type, value}) => {
        switch (type) {
            case 'month': return `<b>${value}</b>`;
            default     : return value;
        }
    })
    .reduce((string, part) => string + part);

console.log(dateFormat.format(date)); // "10/30/2015"
console.log(formattedDate);           // "10/30/2015"
console.log(fancyFormattedDate);      // "<b>10</b>/30/2015"
```

as you can see in https://github.com/tc39/ecma402/issues/30#issuecomment-152446926

If some fields are overlapping, then the result is not an iterable that when joined matches the stringified output.
How should we handle that?

Reply:
> If some fields are overlapping, then the result is not an iterable that when joined matches the stringified output.

True.
The way ICU does it is by using a special buffer where you can append "text with field info"
Can also specify "star of field" and "end of field"
The append adds the plain text to a char buffer, and the field info in separate structure

So the final buffer has the plain text already there, there is no need to create it by joining.

Imagine something like this:
```
{
  plain_text: "Mar 3, 2020, 9:42 pm"
  fields: [
    date: [0, 11],
    time: [13, 20],
    day: [4, 5]
    month: [0, 3]
    year: [7, 11]
    hour: [13, 14]
    min: [15, 17]
    ampm: [18, 20]
  ]
}
```

Reply:
Do you think it helps if I share some code using that API?

To be fair I find the Java [AttributedCharacterIterator](https://docs.oracle.com/javase/8/docs/api/java/text/AttributedCharacterIterator.html) very unfriendly to use.
But was already there, so ICU decided to use it.

Reply:
I've been using FormattedStringBuilder for new classes; all of ICU uses it for fields *except* DateFormat.  FormattedStringBuilder is basically a char array with an equal-length fields array.  The way it handles overlapping fields is by business logic when you request the resolved fields from the FormattedStringBuilder.  The advantage of this approach is that it centralizes the fields logic into one place, rather than having it scattered around the format function.  FormattedStringBuilder is very low-cost if you don't end up using the fields.

Reply:
ECMA-402 does not normally handle overlapping fields.  When it does, it adds them as annotations on existing fields, like the "source" in [formatRangeToParts](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/formatRangeToParts).

Reply:
I think considering the support for overlapping fields from the get go would help in the long run.
I can imagine `MessageFormat` also supporting field info (for placeholders).

So you might end up with "You have an appointment on Tue 12, at 7:30 pm"
With info that "Tue 12, at 7:30 pm" was a placeholder, overlapping a date, and a time, overlapping day of week, day, and so on.

```
You have an appointment on Tue 12, at 7:30 pm
                           [                 )  : event_day_time, MessageFormat.placeholder
                           [     )              : day, DateTimeFormat.field
                           [  )                 : day of week, DateTimeFormat.field
                               [ )              : day of month, DateTimeFormat.field
                                      [      )  : time, DateTimeFormat.field
                                      [)        : hour, DateTimeFormat.field
                                        [ )     : minute, DateTimeFormat.field
                                           [ )  : period, DateTimeFormat.field
```

Reply:
+1 for supporting overlapping fields in the API.

The question on the table right now is how we want to express fields in the internal data model.  There are two competing constraints: expressiveness and performance.  We should think carefully about how to design FormattedStringBuilder (#135) in a way that balance these two constraints.

Reply:
We will follow up in #135.